src/efimon/power/rapl.cpp

```diff
@@ -0,0 +1,170 @@
+#include <efimon/status.hpp>
+#include <fstream>
+#include <string>
+#include <vector>
+
+namespace efimon {
+
+extern uint64_t GetUptime();
+
+RAPLMeterObserver::RAPLMeterObserver(const uint /* pid */,
+                                     const ObserverScope scope,
+                                     const uint64_t interval)
+    : Observer{}, info_{}, valid_{false} {
+  uint64_t type = static_cast<uint64_t>(ObserverType::CPU) |
+                  static_cast<uint64_t>(ObserverType::POWER) |
+                  static_cast<uint64_t>(ObserverType::INTERVAL);
+
+  this->interval_ = interval;
+  this->device_ = info_.GetNumSockets();
+
+  if (ObserverScope::SYSTEM != scope) {
+    throw Status{Status::INVALID_PARAMETER, "Process-scope is not supported"};
+  }
+
+  this->caps_.emplace_back();
+  this->caps_[0].type = type;
+
+  this->Reset();
+  this->Trigger();
+}
+
+Status RAPLMeterObserver::GetSocketConsumption(const uint socket_id) {
+  /* Make the powercap file */
+  std::string energy_file_name = "/sys/class/powercap/intel-rapl:";
+  energy_file_name += std::to_string(socket_id);
+  energy_file_name += "/energy_uj";
+
+  /* Open file */
+  std::ifstream energy_file;
+  energy_file.open(energy_file_name);
+  if (!energy_file.is_open()) {
+    return Status{Status::NOT_FOUND, "The RAPL Interface cannot be opened"};
+  }
+
+  std::string payload_uj;
+  std::getline(energy_file, payload_uj);
+
+  if (!valid_) {
+    /* Read two times */
+    before_socket_meters_.at(socket_id) = std::stof(payload_uj) * 1e-06;
+    after_socket_meters_.at(socket_id) = std::stof(payload_uj) * 1e-06;
+  } else {
+    std::swap(this->before_socket_meters_.at(socket_id),
+              this->after_socket_meters_.at(socket_id));
+    after_socket_meters_.at(socket_id) = std::stof(payload_uj) * 1e-06;
+  }
+
+  return Status{};
+}
+
+Status RAPLMeterObserver::Trigger() {
+  /* Set readings common metadata */
+  auto time = GetUptime();
+  this->readings_.type = static_cast<uint64_t>(ObserverType::CPU) |
+                         static_cast<uint64_t>(ObserverType::POWER);
+  this->readings_.difference = time - this->readings_.timestamp;
+  this->readings_.timestamp = time;
+  this->readings_.overall_power = 0;
+
+  /* Check if the parse is for a single socket */
+  if (this->device_ < static_cast<uint>(info_.GetNumSockets())) {
+    this->GetSocketConsumption(this->device_);
+    this->ParseResults(this->device_);
+    this->valid_ = true;
+    return Status{};
+  }
+
+  /* Get for all sockets */
+  for (int i = 0; i < info_.GetNumSockets(); ++i) {
+    this->GetSocketConsumption(i);
+    this->ParseResults(i);
+  }
+
+  /* Get for all the sockets */
+  this->valid_ = true;
+  return Status{};
+}
+
+void RAPLMeterObserver::ParseResults(const uint socket_id) {
+  float val = this->after_socket_meters_.at(socket_id) -
+              this->before_socket_meters_.at(socket_id);
+  this->readings_.socket_power.at(socket_id) = val;
+  this->readings_.overall_power += val;
+}
```

Comment:
The point here is that we are accounting for energy, not power. We must divide these counters by the time difference to get the average power and cancel out the time component.

Also, there could be an overflow here. We also need to think about it here.